velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,228 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Aggregate.h"
+#include "velox/expression/FunctionSignature.h"
+#include "velox/functions/prestosql/aggregates/AggregateNames.h"
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/functions/prestosql/aggregates/SimpleNumericAggregate.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+/// ChecksumAggregate is the aggregate function to compute checksum aggregate on
+/// a vector.
+/// Checksum will return an order-insensitive checksum of the input vector.
+///
+/// checksum(T)-> varbinary
+class ChecksumAggregate : public exec::Aggregate {
+ public:
+  static const long kPrime64 = 0x9E3779B185EBCA87L;
+
+  explicit ChecksumAggregate(const TypePtr& resultType)
+      : Aggregate(resultType) {}
+
+  void finalize(char** /* groups */, int32_t /* numGroups */) override {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(int64_t);
+  }
+
+  void initializeNewGroups(
+      char** groups,
+      folly::Range<const vector_size_t*> indices) override {
+    for (auto i : indices) {
+      *value<int64_t>(groups[i]) = 0;
+    }
+  }
+
+  void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto* vector = (*result)->as<FlatVector<StringView>>();
+    VELOX_CHECK(vector);
+    vector->resize(numGroups);
+
+    auto* rawValues = vector->mutableRawValues();
+    for (vector_size_t i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      clearNull(group);
+      rawValues[i] = StringView(value<char>(group), sizeof(int64_t));
+    }
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    auto vector = (*result)->as<FlatVector<int64_t>>();
+    VELOX_CHECK(
+        vector,
+        "Unexpected type of the result vector: {}",
+        (*result)->type()->toString());
+    vector->resize(numGroups);
+
+    for (int32_t i = 0; i < numGroups; ++i) {
+      auto group = groups[i];
+      if (isNull(group)) {
+        vector->setNull(i, true);
+      } else {
+        vector->set(i, *value<int64_t>(group));
+      }
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
```

Comment:
Why do we need this check? PrestoHasher takes care of nulls, no? Can't we just use rawHashes for all rows?